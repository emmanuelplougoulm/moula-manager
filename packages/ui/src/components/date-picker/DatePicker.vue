<template>
  <div class="date-picker-container">
    <LabelItem
      :text="label"
      :highlighted="isHighlighted"
      :tooltip-title="tooltipTitle"
      :tooltip-description="tooltipDescription"
    />
    <DatePicker
      v-model:value="selectedValue"
      format="DD/MM/YYYY"
      value-type="format"
      :lang="lang"
      :editable="false"
      data-testid="date-picker"
      @change="onChange"
      @open="onOpen"
      @close="onClose"
      @click="onClick"
    >
      <template #icon-calendar>
        <CalendarToday />
      </template>
    </DatePicker>
    <CaptionItem :text="caption" />
    <ErrorItem :text="error" />
  </div>
</template>

<script setup lang="ts">
import { ref, computed } from 'vue';
import ErrorItem from '../error/ErrorItem.vue';
import CaptionItem from '../caption/CaptionItem.vue';
import LabelItem from '../label/LabelItem.vue';
import CalendarToday from '@material-symbols/svg-600/outlined/calendar_today-fill.svg?component';
import { getErrorText } from '../../utils/commons';
import DatePicker from 'vue-datepicker-next';
import 'vue-datepicker-next/index.css';
import 'vue-datepicker-next/locale/fr.es';
import 'vue-datepicker-next/locale/en.es';

const props = defineProps({
  label: {
    type: String,
    required: true
  },
  modelValue: {
    type: String,
    required: true
  },
  tooltipTitle: {
    type: String,
    default: undefined
  },
  tooltipDescription: {
    type: String,
    default: undefined
  },
  required: {
    type: Boolean,
    default: false
  },
  lang: {
    type: String,
    required: true
  },
  errorText: {
    type: String,
    default: ''
  },
  caption: {
    type: String,
    default: undefined
  }
});

const selectedValue = ref(props.modelValue);
const error = ref(props.errorText);
const opened = ref(false);
const clicked = ref(false);

const isHighlighted = computed(() => {
  return !selectedValue.value && props.required && !clicked.value;
});
const border = computed(() => (isHighlighted.value ? '0.125rem solid var(--primary-30)' : 'none'));
const backgroundColor = computed(() =>
  isHighlighted.value ? 'var(--primary-90)' : 'var(--neutral-80)'
);
const textColor = computed(() => (isHighlighted.value ? 'var(--primary-30)' : 'var(--neutral-20)'));
const svgColor = computed(() => {
  if (opened.value || !isHighlighted.value) {
    return 'var(--neutral-40)';
  } else {
    return 'var(--primary-30)';
  }
});

const emit = defineEmits(['update:modelValue']);
const onChange = (value: any): void => {
  if (value) {
    error.value = '';
    emit('update:modelValue', value);
  } else {
    value = '';
    error.value = getErrorText(props.required, value);
  }
};

const onOpen = () => {
  opened.value = true;
};
const onClick = () => {
  clicked.value = true;
};
const onClose = () => {
  opened.value = false;
  clicked.value = false;
};
</script>

<style>
:is(.mx-input, .mx-input:hover) {
  border: v-bind(border);
}

:is(.mx-icon-calendar svg, .mx-icon-clear svg) {
  color: v-bind(svgColor);
}

.date-picker-container {
  display: flex;
  flex-direction: column;
  gap: 0.5rem;
  width: 10.3125rem;
}

.mx-datepicker {
  width: 10.3125rem;
}

.mx-input {
  font-size: 1.125rem;
  height: 3rem;
  padding-left: 1rem;
  border-radius: 0.25rem;
  color: v-bind(textColor);
  background-color: v-bind(backgroundColor);
  box-shadow: none;
}

.mx-input:focus-visible,
.mx-icon-clear svg:focus-visible {
  background-color: var(--neutral-70);
  border: none;
  color: var(--neutral-10);
}

.mx-calendar-content .cell.active {
  background-color: var(--primary-30);
}

.mx-btn-text:hover,
.mx-table-date .today,
.mx-table-date .today:hover {
  color: var(--primary-30);
}

.mx-icon-calendar,
.mx-icon-clear {
  right: 1rem;
}

.mx-icon-calendar svg,
.mx-icon-clear svg {
  width: 1.6rem;
  height: 1.6rem;
}
</style>
