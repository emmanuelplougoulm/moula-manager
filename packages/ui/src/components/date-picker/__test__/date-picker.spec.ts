import { mount } from '@vue/test-utils';
import DatePicker from '../DatePicker.vue';

const value = '12/07/1998';
const inputSelector = '.mx-input';
const clearSelector = '.mx-icon-clear';

describe('DatePicker tests', () => {
  it('should display the default class by default', () => {
    const wrapper = mount(DatePicker, {
      props: {
        label: 'Label',
        modelValue: value,
        lang: 'fr'
      }
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should display the default class if there is a value, even if the required prop is true', () => {
    const wrapper = mount(DatePicker, {
      props: {
        label: 'Label',
        modelValue: value,
        required: true,
        lang: 'fr'
      }
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should display the highlighted class if there is no initial value and the required prop is true', () => {
    const wrapper = mount(DatePicker, {
      props: {
        label: 'Label',
        modelValue: '',
        required: true,
        lang: 'fr'
      }
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should display the highlighted class and a caption', () => {
    const wrapper = mount(DatePicker, {
      props: {
        label: 'Label',
        modelValue: '',
        required: true,
        lang: 'fr',
        caption: 'This is a caption'
      }
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should display a required error and the highlighted classes if the input is emptied and the required prop is true', async () => {
    const wrapper = mount(DatePicker, {
      props: {
        label: 'Label',
        modelValue: '',
        required: true,
        lang: 'fr'
      }
    });
    const input = wrapper.find(inputSelector);
    await input.trigger('click');
    await input.setValue('12/01/1998');

    const clear = wrapper.find(clearSelector);
    await clear.trigger('click');

    expect(wrapper.html()).toMatchSnapshot();
  });
});
