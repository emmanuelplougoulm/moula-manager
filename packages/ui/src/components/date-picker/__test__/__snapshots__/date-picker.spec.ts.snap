// Vitest Snapshot v1

exports[`DatePicker tests > should display a required error and the highlighted classes if the input is emptied and the required prop is true 1`] = `
"<div class=\\"date-picker-container\\"><span data-v-0704e93f=\\"\\" id=\\"label-item\\" class=\\"text-small-semi-bold label--default label--highlighted\\" data-testid=\\"label-item\\">Label <!--v-if--></span>
  <div class=\\"mx-datepicker\\">
    <div class=\\"mx-input-wrapper\\"><input name=\\"date\\" type=\\"text\\" autocomplete=\\"off\\" class=\\"mx-input\\" readonly=\\"\\"><i class=\\"mx-icon-clear\\"><svg xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 1024 1024\\" width=\\"1em\\" height=\\"1em\\">
          <path d=\\"M810.005 274.005 572.011 512l237.994 237.995-60.01 60.01L512 572.011 274.005 810.005l-60.01-60.01L451.989 512 213.995 274.005l60.01-60.01L512 451.989l237.995-237.994z\\"></path>
        </svg></i><i class=\\"mx-icon-calendar\\"><svg xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"48\\" height=\\"48\\" viewBox=\\"0 96 960 960\\">
          <path d=\\"M185.087 990.131q-32.507 0-55.862-23.356-23.356-23.355-23.356-55.862V301.087q0-32.74 23.356-56.262 23.355-23.521 55.862-23.521H245v-60h74.609v60h320.782v-60H715v60h59.913q32.74 0 56.262 23.521 23.521 23.522 23.521 56.262v609.826q0 32.507-23.521 55.862-23.522 23.356-56.262 23.356H185.087Zm0-79.218h589.826V486H185.087v424.913Z\\"></path>
        </svg></i></div>
    <!--teleport start-->
    <!--teleport end-->
  </div>
  <!--v-if--><span data-v-c1d8433f=\\"\\" id=\\"error-item\\" data-testid=\\"error-item\\" class=\\"text-small-regular\\">Required</span>
</div>"
`;

exports[`DatePicker tests > should display the default class by default 1`] = `
"<div class=\\"date-picker-container\\"><span data-v-0704e93f=\\"\\" id=\\"label-item\\" class=\\"text-small-semi-bold label--default\\" data-testid=\\"label-item\\">Label <!--v-if--></span>
  <div class=\\"mx-datepicker\\">
    <div class=\\"mx-input-wrapper\\"><input name=\\"date\\" type=\\"text\\" autocomplete=\\"off\\" class=\\"mx-input\\" readonly=\\"\\"><i class=\\"mx-icon-clear\\"><svg xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 1024 1024\\" width=\\"1em\\" height=\\"1em\\">
          <path d=\\"M810.005 274.005 572.011 512l237.994 237.995-60.01 60.01L512 572.011 274.005 810.005l-60.01-60.01L451.989 512 213.995 274.005l60.01-60.01L512 451.989l237.995-237.994z\\"></path>
        </svg></i><i class=\\"mx-icon-calendar\\"><svg xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"48\\" height=\\"48\\" viewBox=\\"0 96 960 960\\">
          <path d=\\"M185.087 990.131q-32.507 0-55.862-23.356-23.356-23.355-23.356-55.862V301.087q0-32.74 23.356-56.262 23.355-23.521 55.862-23.521H245v-60h74.609v60h320.782v-60H715v60h59.913q32.74 0 56.262 23.521 23.521 23.522 23.521 56.262v609.826q0 32.507-23.521 55.862-23.522 23.356-56.262 23.356H185.087Zm0-79.218h589.826V486H185.087v424.913Z\\"></path>
        </svg></i></div>
    <!--teleport start-->
    <!--teleport end-->
  </div>
  <!--v-if-->
  <!--v-if-->
</div>"
`;

exports[`DatePicker tests > should display the default class if there is a value, even if the required prop is true 1`] = `
"<div class=\\"date-picker-container\\"><span data-v-0704e93f=\\"\\" id=\\"label-item\\" class=\\"text-small-semi-bold label--default\\" data-testid=\\"label-item\\">Label <!--v-if--></span>
  <div class=\\"mx-datepicker\\">
    <div class=\\"mx-input-wrapper\\"><input name=\\"date\\" type=\\"text\\" autocomplete=\\"off\\" class=\\"mx-input\\" readonly=\\"\\"><i class=\\"mx-icon-clear\\"><svg xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 1024 1024\\" width=\\"1em\\" height=\\"1em\\">
          <path d=\\"M810.005 274.005 572.011 512l237.994 237.995-60.01 60.01L512 572.011 274.005 810.005l-60.01-60.01L451.989 512 213.995 274.005l60.01-60.01L512 451.989l237.995-237.994z\\"></path>
        </svg></i><i class=\\"mx-icon-calendar\\"><svg xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"48\\" height=\\"48\\" viewBox=\\"0 96 960 960\\">
          <path d=\\"M185.087 990.131q-32.507 0-55.862-23.356-23.356-23.355-23.356-55.862V301.087q0-32.74 23.356-56.262 23.355-23.521 55.862-23.521H245v-60h74.609v60h320.782v-60H715v60h59.913q32.74 0 56.262 23.521 23.521 23.522 23.521 56.262v609.826q0 32.507-23.521 55.862-23.522 23.356-56.262 23.356H185.087Zm0-79.218h589.826V486H185.087v424.913Z\\"></path>
        </svg></i></div>
    <!--teleport start-->
    <!--teleport end-->
  </div>
  <!--v-if-->
  <!--v-if-->
</div>"
`;

exports[`DatePicker tests > should display the highlighted class and a caption 1`] = `
"<div class=\\"date-picker-container\\"><span data-v-0704e93f=\\"\\" id=\\"label-item\\" class=\\"text-small-semi-bold label--default label--highlighted\\" data-testid=\\"label-item\\">Label <!--v-if--></span>
  <div class=\\"mx-datepicker\\">
    <div class=\\"mx-input-wrapper\\"><input name=\\"date\\" type=\\"text\\" autocomplete=\\"off\\" class=\\"mx-input\\" readonly=\\"\\">
      <!----><i class=\\"mx-icon-calendar\\"><svg xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"48\\" height=\\"48\\" viewBox=\\"0 96 960 960\\">
          <path d=\\"M185.087 990.131q-32.507 0-55.862-23.356-23.356-23.355-23.356-55.862V301.087q0-32.74 23.356-56.262 23.355-23.521 55.862-23.521H245v-60h74.609v60h320.782v-60H715v60h59.913q32.74 0 56.262 23.521 23.521 23.522 23.521 56.262v609.826q0 32.507-23.521 55.862-23.522 23.356-56.262 23.356H185.087Zm0-79.218h589.826V486H185.087v424.913Z\\"></path>
        </svg></i>
    </div>
    <!--teleport start-->
    <!--teleport end-->
  </div><span data-v-b41f83c4=\\"\\" data-testid=\\"caption-item\\" class=\\"text-small-regular\\">This is a caption</span>
  <!--v-if-->
</div>"
`;

exports[`DatePicker tests > should display the highlighted class if there is no initial value and the required prop is true 1`] = `
"<div class=\\"date-picker-container\\"><span data-v-0704e93f=\\"\\" id=\\"label-item\\" class=\\"text-small-semi-bold label--default label--highlighted\\" data-testid=\\"label-item\\">Label <!--v-if--></span>
  <div class=\\"mx-datepicker\\">
    <div class=\\"mx-input-wrapper\\"><input name=\\"date\\" type=\\"text\\" autocomplete=\\"off\\" class=\\"mx-input\\" readonly=\\"\\">
      <!----><i class=\\"mx-icon-calendar\\"><svg xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"48\\" height=\\"48\\" viewBox=\\"0 96 960 960\\">
          <path d=\\"M185.087 990.131q-32.507 0-55.862-23.356-23.356-23.355-23.356-55.862V301.087q0-32.74 23.356-56.262 23.355-23.521 55.862-23.521H245v-60h74.609v60h320.782v-60H715v60h59.913q32.74 0 56.262 23.521 23.521 23.522 23.521 56.262v609.826q0 32.507-23.521 55.862-23.522 23.356-56.262 23.356H185.087Zm0-79.218h589.826V486H185.087v424.913Z\\"></path>
        </svg></i>
    </div>
    <!--teleport start-->
    <!--teleport end-->
  </div>
  <!--v-if-->
  <!--v-if-->
</div>"
`;
